cmd/gitserver/internal/vcssyncer/instrumented_syncer.go

```diff
@@ -0,0 +1,107 @@
+package vcssyncer
+
+import (
+	"context"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/sourcegraph/sourcegraph/cmd/gitserver/internal/common"
+	"github.com/sourcegraph/sourcegraph/internal/api"
+	"io"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// fetchBuckets are the buckets used for the fetch and clone duration histograms.
+// The buckets range from .005s to 120 minutes.
+var fetchBuckets = append(
+	[]float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10, 20, 30},
+	prometheus.LinearBuckets(60, 5*60, 24)...,
+)
+
+var (
+	metricFetchDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "vcssyncer_fetch_duration_seconds",
+		Help:    "Time taken to fetch a repository",
+		Buckets: fetchBuckets,
+	}, []string{"type", "success"})
+
+	metricCloneDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "vcssyncer_clone_duration_seconds",
+		Help:    "Time taken to clone a repository",
+		Buckets: fetchBuckets,
+	}, []string{"type", "success"})
+
+	metricIsCloneableDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "vcssyncer_is_cloneable_duration_seconds",
+		Help:    "Time taken to check if a repository is cloneable",
+		Buckets: prometheus.DefBuckets,
+	}, []string{"type", "success"})
+)
+
+func newInstrumentedSyncer(syncer VCSSyncer) VCSSyncer {
+	typ := syncer.Type()
+	typ = strings.ToLower(typ)
+	typ = strings.ReplaceAll(typ, " ", "_")
+	typ = strings.ReplaceAll(typ, "-", "_")
+	typ = strings.ReplaceAll(typ, ".", "_")
+
+	return &instrumentedSyncer{
+		base:               syncer,
+		formattedTypeLabel: typ,
+	}
+}
+
+// instrumentedSyncer wraps a VCSSyncer and records metrics for each method call.
+type instrumentedSyncer struct {
+	base               VCSSyncer
+	formattedTypeLabel string
+}
+
+func (i *instrumentedSyncer) Type() string {
+	return i.base.Type()
+}
+
+func (i *instrumentedSyncer) IsCloneable(ctx context.Context, repoName api.RepoName) error {
+	start := time.Now()
+	succeeded := true
+
+	err := i.base.IsCloneable(ctx, repoName)
+	if err != nil {
+		succeeded = false
+	}
+
+	metricIsCloneableDuration.WithLabelValues(i.formattedTypeLabel, strconv.FormatBool(succeeded)).Observe(time.Since(start).Seconds())
+
+	return err
+}
```

Comment:
I wonder if this is a little less err prone to accidentally returning the wrong error? 

```suggestion
func (i *instrumentedSyncer) IsCloneable(ctx context.Context, repoName api.RepoName) (err error) {
	start := time.Now()
	defer func() {
	  	metricIsCloneableDuration.WithLabelValues(i.formattedTypeLabel, strconv.FormatBool(err == nil)).Observe(time.Since(start).Seconds())
	}()
	
	return i.base.IsCloneable(ctx, repoName)
}
```

Reply:
Good thought! Fixed.